### DISCUSSION TITLE
[OSPP-Week2] Check the `kuhn_nfsp` experiment and read papers

### DISCUSSION THREAD

**Discussion by peterchen96** [JuliaReinforcementLearning/ReinforcementLearning.jl#342] [Archive]
@findmyway 

Hi, I am a student of the OSPP project [Implement Multi-Agent Reinforcement Learning Algorithms in Julia](https://summer.iscas.ac.cn/#/org/prodetail/210370190). This summer, my main mission is to implement some `multi-agent RL algorithms`(such as NFSP and PSRO) in `RLZoo` and add relative experiments in `RLExperiment`. Following is my current progress and this week's plan.

### Current Progress

I have implemented the Neural Fictitious Self-play(NFSP) algorithm roughly and test it on the Kuhn Poker game([KuhnPokerEnv](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningEnvironments/src/environments/examples/KuhnPokerEnv.jl)). See my [project](https://github.com/peterchen96/nfsp_demo) for more details.

### This Week's Plan

* Use nash_conv to evaluate the experiment:

   I should add some `prob` function in QBasedpolicy for some specific stages. For now, [nash_conv](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/639717388f/src/ReinforcementLearningZoo/src/algorithms/cfr/nash_conv.jl) is not available for QBasedpolicy.

* Organize current codes and add them to the corresponding directories in `RL.jl`.

**Comment by findmyway** [comment]
Besides, as we mentioned before, the julia version of KuhnPokerEnv may need to implement the `clone` method to calculate nash_conv

**Reply by peterchen96** [reply]
There is a [copy](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/0aa2076423ed6a42db6dd6309ff95bc69d8df5c3/src/ReinforcementLearningBase/src/interface.jl#L404) method for AbstractEnv. Maybe add Base.:(==) and Base.hash methods for AbstractEnv is enough, which just like in [OpenSpielEnv](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/639717388fb41199c98b90406bea76232bc6294d/src/ReinforcementLearningEnvironments/src/environments/3rd_party/structs.jl#L31-L32).

**Comment by peterchen96** [comment]
For now, nash_conv can be used in the [kuhn_nfsp](https://github.com/peterchen96/nfsp_demo/blob/848272378ae87234f64d0f01a7c87e6a92266e93/kuhn_nfsp.jl#L69) experiment, although its result looks bad.

So in this week, I would do the following things:

* Check the process of `kuhn_nfsp` to correct the result of the experiment. For more details about the experiment see https://github.com/peterchen96/nfsp_demo.

* Submit `nfsp` algorithm and `kuhn_nfsp` experiment to `ReinforcementLearning.jl` step by step.

* Read papers about overview for cooperative Multi-agent Reinforcement Learning, such as https://arxiv.org/abs/1908.03963.

**Reply by findmyway** [reply]
You'd better create a draft PR so that I can offer suggestions from my side early, otherwise, things may out of control very quickly.

**Reply by findmyway** [reply]
And I can also help to diagnose the possible problems.

**Reply by peterchen96** [reply]
Ok, I'll create a PR later today.

**Comment by peterchen96** [comment]
Tasking List for  adding `Neural Fictitious Self-play(nfsp)` algorithm  and  `kuhn_nfsp` experiment

### NFSPAgent's structure:

* rl_agent: 

    ```julia
    Agent(
        policy = QBasedPolicy(
            learner = DQNLearner,
            explorer = EpsilonGreedyExplorer,
        ),
        trajectory = CircularArraySARTTrajectory
    )
    ```

    where `rl_agent`(use DQN as an example) works to search for the best response from the self-play process.

* sl_agent:
    
    ```julia
    Agent(
        policy = BehaviorCloningPolicy,
        trajectory = ReservoirTrajectory,
    )
    ```
    
    where `sl_agent` learns the best response for the state from rl_agent's policy.

### Task list

- [X] ~~add `AverageLearner` in `RLZoo`~~ use `BehaviorCloingPolicy` as sl_agent for now
- [X] add `nfsp` and its relative operations 
- [X] add `kuhn_nfsp` experiment

New progress is in #402.
